contracts/governance/extensions/GovernorTimelock.sol

```diff
@@ -0,0 +1,191 @@
+
+    /**
+     * @dev Overridden version of the {Governor-state} function with added support for the `Queued` and `Expired` state.
+     */
+    function state(uint256 proposalId) public view virtual override(IGovernor, Governor) returns (ProposalState) {
+        ProposalState currentState = super.state(proposalId);
+
+        if (currentState == ProposalState.Succeeded && proposalEta(proposalId) != 0) {
+            return ProposalState.Queued;
+        } else {
+            return currentState;
+        }
+    }
+
+    /**
+     * @dev Public accessor to check the eta of a queued proposal.
+     */
+    function proposalEta(uint256 proposalId) public view virtual override returns (uint256) {
+        return _proposalEta[proposalId];
+    }
+
+    /**
+     * @dev Public accessor to check the execution details.
+     */
+    function proposalExecutionDetails(uint256 proposalId) public view virtual returns (ExecutionDetail[] memory) {
+        return _executionDetails[proposalId];
+    }
+
+    /**
+     * @dev See {IGovernor-propose}
+     */
+    function propose(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        string memory description
+    ) public virtual override(IGovernor, Governor) returns (uint256) {
+        uint256 proposalId = super.propose(targets, values, calldatas, description);
+
+        ExecutionDetail[] storage details = _executionDetails[proposalId];
+        for (uint256 i = 0; i < targets.length; ++i) {
+            details.push(_detectExecutionDetails(targets[i], bytes4(calldatas[i])));
+        }
+
+        return proposalId;
+    }
+
+    /**
+     * @dev Function to queue a proposal to the timelock.
+     *
+     * NOTE: execution delay is estimated based on the delay information retreived in {proposal}. This value may be
+     * off if the delay were updated during the vote.
+     */
+    function queue(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        bytes32 descriptionHash
+    ) public virtual override returns (uint256) {
+        uint256 proposalId = hashProposal(targets, values, calldatas, descriptionHash);
+
+        ProposalState currentState = state(proposalId);
+        if (currentState != ProposalState.Succeeded) {
+            revert GovernorUnexpectedProposalState(
+                proposalId,
+                currentState,
+                _encodeStateBitmap(ProposalState.Succeeded)
+            );
+        }
+
+        uint32 setback = 0;
+
+        ExecutionDetail[] memory details = proposalExecutionDetails(proposalId);
+        for (uint256 i = 0; i < targets.length; ++i) {
+            if (details[i].manager != address(0)) {
+                IAccessManager(details[i].manager).schedule(targets[i], calldatas[i]);
+            }
+            setback = uint32(Math.max(setback, details[i].delay)); // cast is safe, both parameters are uint32
+        }
```

Comment:
I see what you mean. I planned to use memory instead of storage when fetching `_executionDetails[proposalId]` because everything will be accessed anyway (and I wanted to avoid double sloads). I did no see the length in the heap as a cost issue, but you're right that be don't need everything to be available in memory at the same time, so we can extend less.